### ISSUE TITLE
Query Frontend: missing port in address

### ISSUE THREAD

**Issue by karlism** [thanos-io/thanos#4118]
**Thanos, Prometheus and Golang version used**:

```
thanos, version 0.20.0 (branch: HEAD, revision: 893188a6752e260cf54f3c082392a521b6c92f75)
  build user:       root@83239bffd915
  build date:       20210428-12:54:57
  go version:       go1.16.3
  platform:         linux/amd64
```

**Object Storage Provider**:
Irrelevant

**What happened**:
After an upgrade from Thanos 0.19.0 to 0.20.0, Thanos Ruler is logging following errors:
```
Apr 29 06:03:56 hostname thanos: level=error ts=2021-04-29T06:03:56.854131338Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address metrics.example.com: missing port in address"
Apr 29 06:03:56 hostname thanos: level=error ts=2021-04-29T06:03:56.854439031Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address metrics.example.com: missing port in address"
```
Following parameters are being used to start the service:
`/usr/bin/thanos rule --data-dir=/var/lib/thanos/rule --grpc-address=0.0.0.0:10910 --http-address=0.0.0.0:10911 --rule-file=/etc/prometheus/thanos_alerts.yml --query=localhost:10913/thanos/query --alertmanagers.url=https://alerts.example.com --alert.query-url=https://metrics.example.com/all --web.external-prefix=/thanos/rule`
If I change `-alert.query-url` parameter value to `https://metrics.example.com:443/all`, these errors are gone.

**What you expected to happen**:
It is perfectly fine for URL not to contain port number when default port is used. I would expect `https://metrics.example.com/all` to work properly.

**How to reproduce it (as minimally and precisely as possible)**:
Run Thanos Ruler service with `--alert.query-url=https://metrics.example.com/all` flag.

**Anything else we need to know**:
No.

**Comment by yeya24** [comment]
I cannot reproduce it by adding `--alert.query-url=https://metrics.example.com/all`.

From the given log:
`Apr 29 06:03:56 hostname thanos: level=error ts=2021-04-29T06:03:56.854131338Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address metrics.example.com: missing port in address"`

The protocol here is `http` instead of `https`. This looks weird and we need more investigation on it.

**Comment by karlism** [comment]
@yeya24, thanks for looking into this!

I've investigated it bit further, and it appears that issue is not with Rule but with Query Frontend. I initially assumed that it was Rule issue as the only place where we reference metrics.example.com server is Rule command line parameters. As for the reason why adding `:443` to URL resolved an issue - we are running everything behind load balancers and load balancer failed over to other backend node, so it appeared that this had solved an issue, while in reality these errors were still there, but on different Thanos node.

I added `SyslogIdentifier` parameter to better distinguish different Thanos service logs:
```
# cat /etc/systemd/system/thanos-query-frontend.service.d/identifier.conf 
[Service]
SyslogIdentifier=thanos-query-frontend
```
Errors are now following, and it is clear that these are logged by Query Frontend service:
```
Apr 30 07:36:24 hostname thanos-query-frontend: level=error ts=2021-04-30T07:36:24.670221963Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address metrics.example.com: missing port in address"
Apr 30 07:36:24 hostname thanos-query-frontend: level=error ts=2021-04-30T07:36:24.884668317Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address metrics.example.com: missing port in address"
Apr 30 07:36:24 hostname thanos-query-frontend: level=error ts=2021-04-30T07:36:24.933126736Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address metrics.example.com: missing port in address"
Apr 30 07:36:25 hostname thanos-query-frontend: level=error ts=2021-04-30T07:36:25.222138567Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address metrics.example.com: missing port in address"
```
I've also tried enabling debug logging, but it didn't produce any additional output.

In our current setup we are always connecting to Thanos services through Nginx load balancer - metrics.example.com. I see these errors logged every time I open Query Frontend web page on load balancer by following URL: `https://metrics.example.com/thanos/query/`

If I connect to Thanos Query Frontend server directly without load balancers, there are no errors in logs: `http://thanos1:10913/thanos/query/`

Relevant NGINX load balancer configuration:
```
    upstream thanos_query_frontend {
        server thanos1:10913;
        server thanos2:10913;
    }
   ...
        location /thanos/query/ {
            proxy_pass http://thanos_query_frontend/thanos/query/;
            proxy_redirect off;
            proxy_set_header X-Real-IP $remote_addr;
            proxy_set_header Host $host;
            proxy_set_header X-Forwarded-For $proxy_add_x_forwarded_for;
            proxy_set_header X-Forwarded-Proto $scheme;
        }
```
Thanos Query Frontend service is running with following flags:
`/usr/bin/thanos query-frontend --http-address=0.0.0.0:10913 --query-range.response-cache-config-file=/etc/prometheus/thanos_query_frontend_cache.yml --query-frontend.downstream-url=http://localhost:10904 --query-frontend.log-queries-longer-than=5s --query-frontend.compress-responses`
And has following config:
```
# cat /etc/prometheus/thanos_query_frontend_cache.yml 
---
#
# Ansible managed
#

type: IN-MEMORY
config:
  max_size: "2GB"
  max_size_items: 0
  validity: 0s
```
Thanos Query service is running on same servers where Thanos Query Frontend, with following flags:
`/usr/bin/thanos query --grpc-address=0.0.0.0:10903 --http-address=0.0.0.0:10904 --query.replica-label=replica --store=prom1a:10901 --store=prom1b:10901 --query.timeout=1m --store.response-timeout=10000ms --web.external-prefix=/thanos/query`

**Comment by karlism** [comment]
As a test, I've changed Nginx `Host` header from `$host` to `kaka.example.com` and now Query Frontend logs contain following error:
```
Apr 30 08:03:16 thanos1 thanos-query-frontend: level=error ts=2021-04-30T08:03:16.516654107Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address kaka.example.com: missing port in address"
```
So it is probably safe to assume that Thanos Query Frontend has some issues when Host header is set.

**Comment by sepich** [comment]
Confirming the issue while running query-frontend v0.20.1 behind ingress-nginx in k8s.
As a workaround have to add annotation
`nginx.ingress.kubernetes.io/upstream-vhost: "$host:10902"`

**Comment by tiagomeireles** [comment]
I'm also experiencing this while running query-frontend behind ingress-nginx. Is this a configuration issue or an application bug?

**Comment by mpritter76** [comment]
I'm also seeing the same issue and in my case running behind istio. query-frontend config is:
```
spec:
  containers:
  - args:
    - query-frontend
    - --log.level=info
    - --log.format=logfmt
    - --log.request.decision=LogFinishCall
    - --http-address=0.0.0.0:10902
    - --query-range.split-interval=24h
    - --query-range.max-retries-per-request=5
    - --query-range.max-query-parallelism=14
    - --query-range.response-cache-max-freshness=1m
    - --query-frontend.downstream-url=http://thanos-query-query-http.kube-system.svc.cluster.local:10902
    - --query-frontend.compress-responses
    - --labels.split-interval=24h
    - --labels.max-retries-per-request=5
    - --labels.max-query-parallelism=14
    - --labels.response-cache-max-freshness=1m
    - --labels.response-cache-max-freshness=1m
    - --query-frontend.log-queries-longer-than=1s
```

**Comment by aeugenio** [comment]
we are also seeing the same error with the query component 

```level=error ts=2021-05-19T00:34:14.714147632Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address tq-xavier.prod-us-east-2-tsdb.moogsoft.cloud: missing port in address"
level=error ts=2021-05-19T00:34:14.718846301Z caller=http.go:48 protocol=http http.component=server msg="failed to parse host port for http log decision" err="address tq-xavier.prod.ourcompany.cloud: missing port in address"```

we have an nginx ingress-controller that uses an NLB in a private subnet.

**Comment by karlism** [comment]
Seeing same behavior with 0.21.0 too:
```
$ thanos --version
thanos, version 0.21.0 (branch: HEAD, revision: d19aea21b178a6c08838c6a05b834dd44ef14273)
  build user:       root@30b80273f41d
  build date:       20210603-13:49:01
  go version:       go1.16.4
  platform:         linux/amd64
```

**Comment by yeya24** [comment]
Should be solved in #4348. Thanks @spaparaju
